sdk/cosmos/azure-cosmos/azure/cosmos/_service_request_retry_policy.py

```diff
@@ -43,28 +45,85 @@ def ShouldRetry(self):
+                    if self.failover_retry_count >= self.total_retries:
+                        return False
+                    self.location_endpoint = self.resolve_next_region_service_endpoint()
 
-            # Resolve the endpoint for the request and pin the resolution to the resolved endpoint
-            # This enables marking the endpoint unavailability on endpoint failover/unreachability
-            self.location_endpoint = self.global_endpoint_manager.resolve_service_endpoint(self.request)
+            print("In service request - Location endpoint: {}".format(self.location_endpoint))
             self.request.route_to_location(self.location_endpoint)
         return True
+
+    # This function prepares the request to go to the second endpoint in the same region
+    def resolve_current_region_service_endpoint(self):
+        # clear previous location-based routing directive
+        self.request.clear_route_to_location()
+        self.in_region_retry_count += 1
+        print("In service request retry policy - Resolving current region endpoint - in region retry count: {}"
+              .format(self.in_region_retry_count))
+        # resolve the next service endpoint in the same region
+        # since we maintain 2 endpoints per region for write operations
+        self.request.route_to_location_with_preferred_location_flag(self.failover_retry_count, True)
+        return self.global_endpoint_manager.resolve_service_endpoint(self.request)
+
+    # This function prepares the request to go to the next region
+    def resolve_next_region_service_endpoint(self):
+        # This acts as an index for next location in the list of available locations
+        # clear previous location-based routing directive
+        print("In service request retry policy - Resolving next region endpoint - failover retry count: {}"
+              .format(self.failover_retry_count))
+        self.request.clear_route_to_location()
+        # clear the last routed endpoint within same region since we are going to a new region now
+        self.request.clear_last_routed_location()
+        # set location-based routing directive based on retry count
+        # ensuring usePreferredLocations is set to True for retry
+        self.request.route_to_location_with_preferred_location_flag(self.failover_retry_count, True)
+        # Resolve the endpoint for the request and pin the resolution to the resolved endpoint
+        # This enables marking the endpoint unavailability on endpoint failover/unreachability
+        return self.global_endpoint_manager.resolve_service_endpoint(self.request)
+
+    def mark_endpoint_unavailable(self, refresh_cache: bool):
+        print("Marking endpoint unavailable - refresh_cache : {}".format(refresh_cache))
+        if _OperationType.IsReadOnlyOperation(self.request.operation_type):
+            self.global_endpoint_manager.mark_endpoint_unavailable_for_read(self.location_endpoint, True)
```

Comment:
nit: log line can precede the actual `mark_endpoint_unavailable_for_*` invocation.

Reply:
Done.